File: runtime/java/src/main/java/dev/argon/verilization/runtime/I32.java
package dev.argon.verilization.runtime;

import java.io.IOException;

public final class I32 {
    private I32() {}

    public static int fromInteger(int i) {
        return i;
    }

    public static final Codec<Integer> codec = new Codec<Integer>() {
        @Override
        public Integer read(FormatReader reader) throws IOException {
            return reader.readInt();
        }

        @Override
        public void write(FormatWriter writer, Integer value) throws IOException {
            writer.writeInt(value);
        }
    };
}

File: runtime/typescript/src/I32.ts
import { Codec } from "./Codec.js";
import { FormatReader, FormatWriter } from "./FormatIO.js";

export type I32 = number;

export const codec: Codec<I32> = {
    async read(reader: FormatReader): Promise<I32> {
        return await reader.readU32() | 0;
    },

    write(writer: FormatWriter, value: I32): Promise<void> {
        return writer.writeU32(value >>> 0);
    },
};

export function fromInteger(n: bigint): I32 {
    return Number(n) | 0;
}
